rust/kernel/cpu.rs

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! CPU utilities for SMP systems.
+//!
+//! Note: This module is a work in progress.
+
+use crate::{bindings, error::code, prelude::*};
+use core::{marker::PhantomData, mem::ManuallyDrop};
+
+/// Represents an SMP CPU by its zero-based kernel processor id.
+///
+/// Cpu values can be safely passed between threads (and across physical CPUs) as they are
+/// merely an integer index representing a CPU. All actual interaction with CPUs is late-
+/// bound.
+pub struct Cpu(i32);
```

Comment:
Why use a signed type if you're going to reject `< 0` values on construction?

Reply:
Good question -- this either deserves a change or a comment. I did it to follow the kernel's convention of passing CPU arguments as int rather than unsigned. In practice no valid id will ever be negative, so it seems like the rusty thing to do here would be to go unsigned and cast as appropriate for the kernel. Is that consistent with the overall approach being taken (if so, happy to swap to u32)?

Reply:
Yes, I think using `u32` would be more consistent with how we're doing things in Rust.